### PR TITLE
Setlist Ownership & Basic Permissions

### LINKED ISSUE #25: Remove Edit and Add Songs on Setlist Page for non-authors

### PR DESCRIPTION
**Pull request by drewlyton**
Adds same concept of `authorId` to `Setlists` 

Closes #25

### CODE REVIEW

File: .env.test [hunk 1]
```diff
@@ -1,2 +1,3 @@
-DATABASE_URL="postgresql://postgres:postgres@localhost:5432/test"
-SESSION_SECRET="thisIsATestSecret1234"
\ No newline at end of file
+DATABASE_URL="file:./data.test.db?connection_limit=1"
```
Comment (drewlyton) [inline .env.test:1]: DB for running tests

File: app/db/setlist.db.ts [hunk 2]
```diff
@@ -23,8 +23,21 @@ export async function getSetlists() {
   });
 }
 
+export async function getSetlistsByUserId(userId: number) {
```
Comment (drewlyton) [inline app/db/setlist.db.ts:26]: Copy pasta from `getSongsByUserId` but adds the include logic for songs

File: app/routes/__app/setlists/$id/index.tsx [hunk 3]
```diff
@@ -15,21 +15,23 @@ import { typedjson, useTypedLoaderData } from "remix-typedjson";
 import { Button } from "~/components/Button";
 import { Stanza } from "~/components/Stanza";
 import { getSetlist } from "~/db/setlist.db";
+import { currentAuthedUser } from "~/utils/auth.server";
 
-export async function loader({ params }: LoaderArgs) {
+export async function loader({ params, request }: LoaderArgs) {
   if (!params.id) throw new Response("Not Found", { status: 404 });
-
+  const user = await currentAuthedUser(request);
```
Comment (drewlyton) [inline app/routes/__app/setlists/$id/index.tsx:22]: Get the current `User` so we know whether they are the `Setlist` author

File: app/routes/__app/setlists/$id/index.tsx [hunk 4]
```diff
@@ -63,17 +65,19 @@ export default function ViewSetlist() {
...
-          className="absolute top-2 right-4 flex items-center text-xs"
-          size="sm"
-          variant="text"
-        >
-          <span className="mr-2">
-            <PencilSquareIcon width={"1em"} />
-          </span>
-          Edit Details
-        </Button.Link>
+        {isUserAuthor && (
```
Comment (drewlyton) [inline app/routes/__app/setlists/$id/index.tsx:68]: Only show `Edit Details` button if they are the author

File: app/routes/__app/setlists/$id/index.tsx [hunk 5]
```diff
@@ -87,7 +91,7 @@ export default function ViewSetlist() {
       </div>
       <div className="flex">
         <h2 className="flex-1 text-2xl">Songs</h2>
-        {!!setlist.songs.length && (
+        {isUserAuthor && !!setlist.songs.length && (
```
Comment (drewlyton) [inline app/routes/__app/setlists/$id/index.tsx:94]: Only show `Add Songs` button if they are the author

File: app/routes/__app/setlists/$id/index.tsx [hunk 6]
```diff
@@ -133,6 +137,7 @@ export default function ViewSetlist() {
               song={song}
               order={setlist.songs.length - index}
               removeSong={removeSong}
+              isEditable={isUserAuthor}
```
Comment (drewlyton) [inline app/routes/__app/setlists/$id/index.tsx:140]: Add `isEditable` flag to hide menu on Song card if not author

File: app/routes/__app/setlists/index.tsx [hunk 7]
```diff
@@ -1,14 +1,16 @@
...
-import { getSetlists } from "~/db/setlist.db";
+import { getSetlistsByUserId } from "~/db/setlist.db";
 import { useSearch } from "~/lib/useSearch";
+import { requireAuthentication } from "~/utils/auth.server";
 
-export async function loader() {
-  const setlists = await getSetlists();
+export async function loader({ request }: LoaderArgs) {
+  const user = await requireAuthentication(request);
+  const setlists = await getSetlistsByUserId(user.id);
```
Comment (drewlyton) [inline app/routes/__app/setlists/index.tsx:13]: Only load `Setlists` the `User` has authored on the Setlist page

File: app/routes/__app/songs/new.tsx [hunk 8]
```diff
@@ -1,10 +1,10 @@
 import type { LoaderArgs } from "@remix-run/node";
 import { redirect } from "@remix-run/node";
 import { createSong } from "~/db/song.db";
-import { currentAuthedUser } from "~/utils/auth.server";
+import { requireAuthentication } from "~/utils/auth.server";
 
 export async function action({ request }: LoaderArgs) {
-  const user = await currentAuthedUser(request);
+  const user = await requireAuthentication(request);
```
Comment (drewlyton) [inline app/routes/__app/songs/new.tsx:7]: Require auth to make a new song

**Review by kstolte — APPROVED** [review]

**Pull request merged**